### ISSUE TITLE
Array's Pop functions reference leak

### ISSUE THREAD

**Issue by felaugmar** [godotengine/godot#77630]
### Godot version

v4.0.2.stable.official [7a0977ce2]

### System information

Windows 10

### Issue description

Array's `pop_at`, `pop_front` and `pop_back` removes an element from the array but messes with the reference count, leaking references.

### Steps to reproduce

Copy, paste and run the snippet below, check the results:

```gdscript
extends Node


func _ready() -> void:
  var ref := RefCounted.new()
  var ref_count : int
  var ref_arr : Array[RefCounted] = []

  ref_arr.resize(10)
  ref_arr.fill(ref)

  ref_count = ref.get_reference_count()
  print('Expected: 11, Result: %d' % ref_count)

  ref_arr.remove_at(0)

  ref_count = ref.get_reference_count()
  print('Expected: 10, Result: %d' % ref_count)

  ref_arr.pop_at(0) # FAIL
  #ref_arr.pop_front() # FAIL
  #ref_arr.pop_back() # FAIL
  #ref_arr.remove_at(0) # OK
  #ref_arr.resize(8) # OK

  ref_count = ref.get_reference_count()
  print('Expected: 9, Result: %d' % ref_count)

```

Output example for `pop_at`, `pop_front` and `pop_back`:
```
Expected: 11, Result: 11
Expected: 10, Result: 10
Expected: 9, Result: 10
```

Output example for `remove_at` and `resize`:
```
Expected: 11, Result: 11
Expected: 10, Result: 10
Expected: 9, Result: 9
```

### Minimal reproduction project

N/A

**Comment by bitsawer** [comment]
Current master gives the right results:

```
Expected: 11, Result: 11
Expected: 10, Result: 10
Expected: 9, Result: 9
```

There has been some fixes in how GDScript handles temporary reference lifetimes. As pop_x() methods return a temporary reference, that might be the cause.

**Comment by AThousandShips** [comment]
@felaugmar can you test it on 4.0.3?

**Comment by felaugmar** [comment]
> @felaugmar can you test it on 4.0.3?

The issue persists.

**Comment by AThousandShips** [comment]
Then hopefully some bug fix can be cheerry picked for 4.0.4

**Comment by bitsawer** [comment]
Also, these dangling temporary references are still freed when the function scope ends, so they don't cause leaks in a traditional sense, they just live a bit longer than expected. Assuming this is the case I'm thinking about.

**Comment by felaugmar** [comment]
By the way, simulating the `pop_*` functions with `get` and `remove_at` works as expected.

**Comment by bitsawer** [comment]
@felaugmar is this issue still valid in latest 4.1? Like I mentioned earlier in this thread, some of these issues were already fixed in 4.1 dev back then, how is the situation now with Godot 4.1.1 stable?
